### ISSUE TITLE
[12.0] point of sale: ticket with zero total amount has non-zero taxes

### ISSUE THREAD

**Issue by dalonsod** [odoo/odoo#55114]
Impacted versions: 12.0 at least

Steps to reproduce:
- define a 21% tax included in Price.
- define some price products applying the above tax.
- for the selected shop, set "Tax-included price" for the Product Prices setting.
- create a POS ticket with some products with the tax above, some of them refunded, forcing a zero total combining them.

Current behavior: due to rounding criteria, total base amount + taxes amount always equals to total amount, but in some corner cases getting a zero total amount it's possible to get with non-sense values (an X base and -X taxes, with X<>0), selecting the appropiated prices and quantities.

Expected behavior: when total amount is zero, both base and taxes should be exactly 0 too, maybe with a workaround for these corner cases.

Video/Screenshot link (optional): configured and tested on latest v12 runbot

A ticket with some products with 21% tax that sums 0 with non-zero taxes, POS view:

![imagen](https://user-images.githubusercontent.com/12749832/88794073-28eed480-d19e-11ea-9d09-ccd4c3850326.png)

Ticket receipt:

![imagen](https://user-images.githubusercontent.com/12749832/88794107-3b690e00-d19e-11ea-95a1-e7dee6fa890a.png)

Ticket data from backend view:

![imagen](https://user-images.githubusercontent.com/12749832/88794239-7408e780-d19e-11ea-9a37-090032fa7970.png)

**Comment by dalonsod** [comment]
@legalsylvain @chienandalu what do you think about this issue? What should be the best approach to solve it? Thanks!!

**Comment by legalsylvain** [comment]
Cc @houssine78 : is it a problem you still faced ?

**Comment by houssine78** [comment]
@legalsylvain no AFAIR I never faced this problem.

**Comment by chienandalu** [comment]
@dalonsod is your rounding setting global or lineal?

**Comment by dalonsod** [comment]
>  is a total vat excl correct ? If yes, we could overwrite vat value with vat incl - vat excl. What do you think ?

in backend are available `amount_total` (vat included) and `amount_tax` (vat value), which is the incorrect one. Total vat excluded is not stored. So, we cannot use that formula IMO.

However, in POS frontend, this calculation is made in a different way: the total vat included is obtained adding the previously calculated total vat excluded and vat value. The result is actually the same, since in this case both total vat excluded and vat value are wrong, obtaining a correct total vat included.

So, in both cases vat value is wrong.

**Comment by dalonsod** [comment]
> 
> 
> @dalonsod is your rounding setting global or lineal?

It was linear rounding. Anyway, I've tested with global rounding and seems to be worse: got 0 taxes... but non zero total amount :(

**Comment by chienandalu** [comment]
Yes, I asked you because we had some issues related with taxes included in price and rounding in general, and the final conclussion was that we should avoid taxes included in price in general and when we had to use linear rounding.

Open an OPW if you have the chance to enforce this issue.

@pedrobaeza Do you know who to ping to?

**Comment by pedrobaeza** [comment]
Maybe @pimodoo can tell us about the problem.
